common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp

```diff
@@ -647,8 +945,175 @@ bool ap_wlan_hal_dwpal::sta_bss_steer(const std::string &mac, const std::string
+    // a BSS on the corresponding radio for each of the M2. If the Multi-AP Agent is currently
+    // operating a BSS with operating parameters that do not completely match any of the M2 in
+    // the received AP-Autoconfiguration WSC message, it shall tear down that BSS.
+
+    // Start by disabling all the existent VAPs
+    for (auto &it : hostapd_config_vaps) {
+        const std::string &vap_if                    = it.first;
+        std::vector<std::string> &vap_hostapd_config = it.second;
+
+        if (!hostapd_config_set_value(vap_hostapd_config, "start_disabled", "1")) {
+            LOG(ERROR) << "Autoconfiguration: cannot disable VAP " << vap_if << ", ignoring";
+        }
+    }
+
+    // Find first VAP entry
+    auto hostapd_vap_iterator = hostapd_config_vaps.begin();
+
+    // Go through the bss_info_conf_list and change the hostapd config accordingly
+    for (const auto &bss_info_conf : bss_info_conf_list) {
+        // If a Multi-AP Agent receives an AP-Autoconfiguration WSC message containing an M2 with
+        // a Multi-AP Extension subelement with bit 4 (Tear Down bit) of the subelement value set
+        // to one (see Table 4), it shall tear down all currently operating BSS(s) on the radio
+        // indicated by the Radio Unique Identifier, and shall ignore the other attributes in the M2.
+        if ((bss_info_conf.bss_type & WSC::eWscVendorExtSubelementBssType::TEARDOWN) != 0) {
+            LOG(DEBUG) << "Autoconfiguration: received teardown, continuing";
+            continue;
+        }
+
+        auto auth_type =
+            son::wireless_utils::wsc_to_bwl_authentication(bss_info_conf.authentication_type);
+        if (auth_type == "INVALID") {
+            LOG(ERROR) << "Autoconfiguration: invalid auth_type "
+                       << int(bss_info_conf.authentication_type);
+            return false;
+        }
+        auto enc_type = son::wireless_utils::wsc_to_bwl_encryption(bss_info_conf.encryption_type);
+        if (enc_type == "INVALID") {
+            LOG(ERROR) << "Autoconfiguration: invalid enc_type "
+                       << int(bss_info_conf.encryption_type);
+            return false;
+        }
+        auto bss_type = son::wireless_utils::wsc_to_bwl_bss_type(bss_info_conf.bss_type);
+        if (bss_type == beerocks::BSS_TYPE_INVALID) {
+            LOG(ERROR) << "Autoconfiguration: invalid bss_type";
+            return false;
+        }
+        LOG(DEBUG) << "Autoconfiguration for ssid: " << bss_info_conf.ssid
+                   << " auth_type: " << auth_type << " encr_type: " << enc_type
+                   << " network_key: " << bss_info_conf.network_key << " bss_type: " << bss_type;
+
+        // We cannot configure STAs yet
+        if ((bss_info_conf.bss_type & WSC::eWscVendorExtSubelementBssType::BACKHAUL_STA) != 0) {
+            LOG(ERROR) << "Autoconfiguration: STAs are not yet supported";
+            return false;
+        }
```

Comment:
Yes, it's not covered in the spec, but I have a suspicion it will be needed unless the expectation is that people should run around pushing buttons every time the backhaul SSID changes. If it ends up to be needed then its place is here, if not then in the son_slave_thread. I'd prefer to avoid moving things around unless it's useful. Do you have access to R2 or 3 of the spec (I only have R1)?

Reply:
I have access to R2, and they haven't clarified this.

Using this to implement backhaul STA is a very liberal interpretation of the spec I think. The problem is that the backhaul STA is not tied to a particular radio, while the autoconfig messages are tied to a radio. If you have a radio dedicated to backhaul STA, it wouldn't even do any autoconfig (because you don't want VAPs on it).

Anyway, if it is used for backhaul STA configuration, then this is *definitely* not the place to handle it. STA handling is in sta_wlan_hal, and it's not talking to hostapd, it's talking to wpa_supplicant. And it's done in the backhaul manager, not in the slave.

Reply:
Note that this is a nitpick.

Reply:
Ok, at the moment wpa_supplicant is not even built into the RAX40 image, and its possible to configure a virtual interface into a client mode, which results in the hostapd configuration "mode" for for that interface changing. Perhaps it's something specific to the Intel driver (i.e. when the supplicant functionality is built-in and the external supplicant is not required). 
I don't know the details. I'll just change the messaging to say we cannot configure STA here and leave it in place. If it changes the message will go away, if not it won't hurt to have it here.